docs/proposal-001_add-ons.md

```diff
@@ -0,0 +1,151 @@
+# Desing Proposal #001: Add-ons
+
+## What's an add-ons?
+
+An add-on extends functionality of a Kubernetes cluster, it may consist of a workload and/or configuration within the give cluster or the cloud provider. The workload, if present, would may classify as an operator or a controller, however that is not neccessary.
+
+The difference between add-ons and anything a user decided to run themselves, is that from user's point of view, add-on is something they don't need to maintain directly. A cluster provider or bootstrap tool (i.e. eksctl), is expected to ensure ease of use, by providing minumum viable (that configuration to ensure end-to-end integration and compatibility), as well as catering for reconfiguration and upgrades.
+
+## Design Requirements and Goals
+
+Add-on must expose all of the following attributes:
+
+- name (fixed) *[TBD]*
+- version tag or revision (has default value that user can override)
+- source URL (has default value that user can override)
+- set of customisations (no default value, user defined)
+
+Internally, an add-on may consist of any and/or all of the following elements:
+
+- Kubernetes workload definitions
+- Kubernetes configuration objects - PersistentVolumes, ConfigMaps, ...etc
+- cloud provider resource definitons (e.g. CloudFormation sub-resources)
+- dependencies (supported Kubernetes versions and corresponding cloud providers, list of references to other add-ons)
+- incompatibilities (supported Kubernetes versions and corresponding cloud providers, list of reference to other add-ons)
+
+In other words, it's ought to be possible to define an add-on that only includes:
+
+- dependencies on other add-ons
+- cloud provider resource definitions
+- Kubernetes configuration objects
+- Kubernetes workload definitions
+
+The implementation must also ensure:
+
+- fully deterministic behaviour for a given version of eksctl, thereby it should have
+    - strict versioning of built-in add-ons by default
+    - any externally-sourced config has to be pinned down or vendored *[TBD]*
+    - there should be an option to use latest version or altenative source URL
```

Comment:
If we allow using the latest version (from a remote location) then do we need to verify that the addon is from the same author and hasn't been tampered with?

Reply:
Would latest come from a eksctl curated repo though?

Reply:
TBD really. What I'm thinking is this:

- eksctl repo has add-on that's sourced from a helm chart and a specific version is vendor
- user may wish to upgrade to latest (or other) version, there must be a way to pull and install that helm chart without too much dancing around

Similar is ought to be possible for non-helm cases.

Reply:
I think this would need a big big banner explaining that it may not work; that it's
* untested
* not signed
* may not work (I can see a lot of the addon charts / CF having to be vendored and altered in some way to work, meaning upstream will not)